src/doc/performance-tuning.md

```diff
@@ -110,95 +205,184 @@ Node 1 HugePages_Total:    32
+memory might be "closer" to one node than another.  The `--cpu`
+argument to a nework function, described above, will issue a warning
+if you use a PCI device that is not local to the NUMA node that
+corresponds to the chosen CPU.
 
-## CPU Isolation/Pinning
-Snabb makes 100% use of a single CPU core, hence its important that no other task ever uses that core. Thats best achieved by telling the kernel scheduler via kernel option 'isolcpus' to ignore cores designated to Snabb. In /etc/default/grub (example reserves cores 18 and 19 for Snabb):
+To determine what PCI devices are local to what NUMA nodes, you need to
+grovel around in `/sys`.  For example if you are going to be working
+with NICs `0000:01:00.0`, `0000:01:00.1`, `0000:02:00.0`, and
+`0000:02:00.1`, check:
 
+```bash
+$ for device in 0000:0{1,2}:00.{0,1}; do \
+    echo $device; cat /sys/bus/pci/devices/$device/numa_node; \
+  done
+0000:01:00.0
+0
+0000:01:00.1
+0
+0000:02:00.0
+0
+0000:02:00.1
+0
 ```
-GRUB_CMDLINE_LINUX_DEFAULT="... isolcpus=18-19 ..."
-```
-Note: Never use CPU 0 for Snabb, because the Linux kernel uses CPU 0 to handle interrupts, including NMI (non-maskable interrupts).
 
-Launch Snabb either via 'taskset' or 'numactl'. Examples to ping it on CPU 18:
+So all of these are on NUMA node 0.  Then you can check your CPUs:
 
 ```
-taskset -c 18 ./snabb ...
-numactl --physcpubind=18 ./snabb ...
+$ numactl -H | grep cpus
+node 0 cpus: 0 1 2 3 4 5
+node 1 cpus: 6 7 8 9 10 11
 ```
 
-Note: Always use 'numactl' on NUMA servers, to limit allocation of memory to specified NUMA nodes with option '--membind=nodes':
+So for these we should run our binaries under `--cpu CPU` to bind them
+to CPUs in the NUMA node 0, and to arrange to use only memory that is
+local to that CPU.
 
-```
-numactl --physcpubind=18 --membind=1  ./snabb ...
-```
-
-## NUMA
-Non-uniform memory access (NUMA) enabled systems have two or more CPU sockets (also called nodes), each with its own memory. While accessing memory across sockets is possible (and happening all the time), its slower than accessing local memory. PCI slots are hard wired to a specific node. It is imperative to pin Snabb to the same node )CPU and memory) as the NIC. Linux offers the command 'lstopo' to get an overall picture in text and graphical form:
+Note that Linux also offers the command `lstopo` to get an overall
+picture in text and graphical form:
 
 ```
 lstopo --of pdf > lstopo.pdf
 ```
 
-Example from a Lenovo RD650 (Intel(R) Xeon(R) CPU E5-2650 v3): 
+Example from a Lenovo RD650 (Intel(R) Xeon(R) CPU E5-2650 v3):
 ![lstopo.png](lstopo.png)
 
-### PCI Card and Snabb on same NUMA node
-Memory shared between Snabb (by means of huge page mapping) and the NIC must share the same node. This is achieved via 'numactl', once the correct node is identified for a given NIC port/PCI address. 
-To find the node for a given PCI address, use cpulistaffinity combined with numactl (TODO: is there a more direct way??):
```

Comment:
Is it worth documenting cpulistaffinity+numactl, in addition to the grovel around in /sys approach?